### PR TITLE
implemented abstract class and two interfaces

### PR DESCRIPTION
**Pull request by lavrivskyi**

### CODE REVIEW

File: src/main/java/core/basesyntax/Circle.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package core.basesyntax;
+
+public class Circle extends Figure {
+    private static final double NUMBER_PI = 3.14;
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/Circle.java:4]: there is a `Math.PI`, more accurate value ;)

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class ColorSupplier {
+    private String[] colors = new String[] {"blue", "red", "green", "ginger", "purple"};
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/ColorSupplier.java:6]: let's create enum `Color`

File: src/main/java/core/basesyntax/FigureDrawer.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax;
+
+public interface FigureDrawer {
+    void drawFigure();
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureDrawer.java:4]:
```suggestion
    void draw();
```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        int randomFigureIndex = random.nextInt(NUMBER_OF_FIGURES);
+        ColorSupplier colorSupplier = new ColorSupplier();
+        String randomColor = colorSupplier.getRandomColor();
+        int randomLength = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE;
+
+        switch (randomFigureIndex) {
+            case 0:
+                return new IsoscelesTrapezoid(randomLength,
+                        randomLength + random.nextInt(MAX_VALUE),
+                        randomLength + random.nextInt(MAX_VALUE), randomColor);
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureSupplier.java:21]: let's create private methods for figure creation and pass variables, not expressions in constructor

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+public class FigureSupplier {
+    private static final int MIN_VALUE = 1;
+    private static final int MAX_VALUE = 127;
+    private static final int NUMBER_OF_FIGURES = 5;
+    private Random random = new Random();
+
+    public Figure getRandomFigure() {
+        int randomFigureIndex = random.nextInt(NUMBER_OF_FIGURES);
+        ColorSupplier colorSupplier = new ColorSupplier();
+        String randomColor = colorSupplier.getRandomColor();
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureSupplier.java:14]: all figures will be of the same color?

File: src/main/java/core/basesyntax/Main.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+public class Main {
+    public static void main(String[] args) {
+        Figure[] figures = new Figure[5];
+        FigureSupplier figureSupplier = new FigureSupplier();
+
+        for (int i = 0; i < figures.length; i++) {
+            figures[i] = figureSupplier.getRandomFigure();
+        }
+
+        for (Figure figure: figures) {
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/Main.java:12]: remove empty lines here

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/FigureSupplier.java (hunk 5 above)
Comment (lavrivskyi) [reply src/main/java/core/basesyntax/FigureSupplier.java]: no, figures will have different colours.

File: src/main/java/core/basesyntax/Colors.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax;
+
+public enum Colors {
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/Colors.java:3]:
```suggestion
public enum Color {
```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    private Random random = new Random();
+
+    public Figure getRandomFigure() {
+        int randomFigureIndex = random.nextInt(NUMBER_OF_FIGURES);
+        ColorSupplier colorSupplier = new ColorSupplier();
+        Colors randomColor = colorSupplier.getRandomColor();
+        int firstRandomNum = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE;
+        int secondRandomNum = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE
+                + random.nextInt(MAX_VALUE);
+        int thirdRandomNum = firstRandomNum + secondRandomNum;
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureSupplier.java:18]: if it's a circle, we'll need only one of these `randomNum`, but we still will create two more. let's create private methods, shall we?

Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureSupplier.java:18]: if it's a circle, we'll need only one of these `randomNum`, but we still will create two more. let's create private methods, shall we?

File: src/main/java/core/basesyntax/Main.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+public class Main {
+    public static void main(String[] args) {
+        Figure[] figures = new Figure[5];
+        FigureSupplier figureSupplier = new FigureSupplier();
+        for (int i = 0; i < figures.length; i++) {
+            figures[i] = figureSupplier.getRandomFigure();
+        }
+        for (Figure figure: figures) {
+            figure.draw();
+        }
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/Main.java:12]: let's make this in one loop

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]
👍

File: src/main/java/core/basesyntax/Main.java (hunk 9 above)
Comment (Max-Pochepets) [reply src/main/java/core/basesyntax/Main.java:12]: upper comment as well ⬆️

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 10]
```diff
@@ -7,15 +7,15 @@
     private static final int MAX_VALUE = 127;
     private static final int NUMBER_OF_FIGURES = 5;
     private Random random = new Random();
+    private int firstRandomNum = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE;
+    private int secondRandomNum = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE
+            + random.nextInt(MAX_VALUE);
+    private int thirdRandomNum = firstRandomNum + secondRandomNum;
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureSupplier.java:13]: now all of these always stored from the creation of the object. why don't we create private methods where these numbers will be initialized?

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 11]
```diff
@@ -7,28 +7,28 @@
     private static final int MAX_VALUE = 127;
     private static final int NUMBER_OF_FIGURES = 5;
     private Random random = new Random();
-    private int firstRandomNum = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE;
-    private int secondRandomNum = random.nextInt((MAX_VALUE - MIN_VALUE) + 1) + MIN_VALUE
-            + random.nextInt(MAX_VALUE);
-    private int thirdRandomNum = firstRandomNum + secondRandomNum;
+
+    private int getRandomNum() {
```
Comment (Max-Pochepets) [inline src/main/java/core/basesyntax/FigureSupplier.java:11]: private methods should be listed after public

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/FigureSupplier.java (hunk 11 above)
Comment (lavrivskyi) [reply src/main/java/core/basesyntax/FigureSupplier.java]: fixed

**Review by Max-Pochepets — APPROVED** [review]

**Pull request closed without merging**